### PR TITLE
Event/alarm framework HLD update. Integrate with event producer framework.

### PR DESCRIPTION
**Pull request by bhaveshdell**
The event/alarm framework HLD was reviewed and committed earlier. 
With the availability of event producer framework, the design is updated to provide DB/persistent support for event history.
Also gnmi/rest openconfig interface for get operations and gnmi subscription.

### CODE REVIEW

**Comment by bhaveshdell** [comment]
@zhangyanzhao Please assign reviewers.

**Comment by bhaveshdell** [comment]
@kwangsuk,  Please review.

**Comment by bhaveshdell** [comment]
@renukamanavalan  Please review update to the existing event framework HLD. 
It is about integrating with  the  producer framework, to provide DB support.

File: doc/event-alarm-framework/event-alarm-framework.md [hunk 1]
```diff
@@ -322,72 +300,22 @@ Developers of new events or alarms need to update this file by declaring name an
...
 
-For alarms:
-```
-    if (temperature >= THRESHOLD) {
-        LOG_EVENT(TEMPERATURE_EXCEEDED, sensor_name_p, RAISE_ALARM, "Temperature for sensor %s is %d degrees", sensor_name_p, current_temp);
-    } else {
-        LOG_EVENT(TEMPERATURE_EXCEEDED, sensor_name_p, CLEAR_ALARM, "Temperature for the sensor %s is %d degrees ", sensor_name_p, current_temp);
-    }
-```
 #### 3.1.1.2 Development Process
```
Comment (jeff-yin) [inline doc/event-alarm-framework/event-alarm-framework.md:304]: It seems like this is introducing differences in the way that an application/component needs to implement their events and alarms. 
If this is the case, we should clearly outline what these changes are, ideally prepare an example, and raise this in the Community during one of the Tuesday calls. 
If you agree, we can work with @zhangyanzhao to schedule something.

**Comment by jeff-yin** [comment]
> The event/alarm framework HLD was reviewed and committed earlier. With the availability of event producer framework, the design is updated to provide DB/persistent support for event history. Also gnmi/rest openconfig interface for get operations and gnmi subscription.

To add some more background, the original event/alarm framework HLD was raised as https://github.com/sonic-net/SONiC/pull/1064 and then approved and merged.
That HLD also had code PRs attached, but the code was never reviewed and merged. 

https://github.com/sonic-net/SONiC/pull/954 for Event Producer was merged and its code PRs were merged, so the events/alarm framework HLD needed to be updated, hence this PR.

File: doc/event-alarm-framework/event-alarm-framework.md (hunk 1 above)
Comment (jeff-yin) [reply doc/event-alarm-framework/event-alarm-framework.md:304]: Resolved.

**Comment by adyeung** [comment]
@rajendra-dendukuri pls help review the HLD

File: doc/event-alarm-framework/event-alarm-framework.md [hunk 2]
```diff
@@ -175,14 +167,11 @@ This modified file can then be uploaded to the device to /etc/evprofile/.
 Operator can select any of these custom event profiles to change default properties of events.
```
Comment (kwangsuk) [inline doc/event-alarm-framework/event-alarm-framework.md:167]: Event profiling to change properties lie severity etc. would be performed by controllers as per their OAM policy.  Hence, can we remove the support to change the profile settings?

**Review by mint570 — APPROVED** [review]

**Comment by zhangyanzhao** [comment]
Reviewed in community on 9/21/2023

**Comment by zhangyanzhao** [comment]
community review recording https://zoom.us/rec/share/P44xWGduoa6Q2Q6D8lNiWyoQlWwlyx5NB2TzgT6QjOaL5ky9QdbwYvAA8tk-iNEh.XPqm29vRYgCiDMz8

**Review by kwangsuk — APPROVED** [review]

**Review by jeff-yin — APPROVED** [review]

**Comment by jeff-yin** [comment]
@zhangyanzhao  can we merge this HLD now?

File: doc/event-alarm-framework/event-alarm-framework.md [hunk 3]
```diff
@@ -102,7 +94,7 @@ Such a change has an important metric called *severity* to indicate how critical
    The set of alarms and their severities are an indication to health of various applications of the system and System LED can be deduced from alarms.
    An acknowledged alarm means that operator is aware of the condition so, acknowledged alarm will be taken out of consideration.
 
-Both events and alarms get recorded in a new DB called EVENT DB in a new redis instance.
+Both events and alarms get recorded in redis DB.
```
Comment (shyam77git) [inline doc/event-alarm-framework/event-alarm-framework.md:102]: which redisDB? could you update if its some existing redisDB or a new one?

Also, as of now, when applications (processes/daemons) publish their event(s)/alarm(s), they do NOT get stored in any redis (or persistent DB) - right? In that case, are they stored by eventD framework/container in memory? no EVENT DB as of now?

File: doc/event-alarm-framework/event-alarm-framework.md [hunk 4]
```diff
@@ -191,65 +180,58 @@ Application owners need to identify various conditions that would be of interest
...
-| 8.6   | event profile   [ default \| name-of-file ]	                                  |                     |
-| 9     | gNMI subscription                                                               |                     |
-| 9.1   | Subscribe to openconfig Event container and Alarm container. All events and alarms published to gNMI subscribed clients. |                    |
-| 10    | Clear all events                                                                |                     |
-| 11    | Any change in open source should be aligned and upstream.                       |                     |
+| 5.2.4 | Openconfig interface  to acknowledge an alarm.                                  |                     |
+| 6     | CLI commands                                                                    |                     |
+| 6.1   | show alarm [ detail \| summary \| severity \| timestamp <from> <to> \| recent <5min\|1hr\|1day> \| sequence-number <from> <to> \| all]          |                    |
+| 6.2   | show event [ detail \| summary \| severity \| timestamp <from> <to> \| recent <5min\|1hr\|1day> \| sequence-number <from> <to>]                 |                     |
```
Comment (shyam77git) [inline doc/event-alarm-framework/event-alarm-framework.md:211]: In 202305 present codebase, do not see support for these CLIs (i.e. show alarm, show event).
[1] In their absence, how does one check eventD's cache-services contents? as to what events, alarms are received and stored in eventD's cache-services?
Is it there as part of some changeset not merged yet? or not there at all?

https://github.com/sonic-net/SONiC/pull/1064 has some unmerged PRs. Does this support reside there?
If so, what's the plan to commit, merge these pending PRs to sonic mainstream repositories?

Comment (bhaveshdell) [reply doc/event-alarm-framework/event-alarm-framework.md:211]: The present codebase does not have CLI support. It supports event publish via telemetry streaming.
The implementation for doc/event-alarm-framework/event-alarm-framework.md, is not merged. 
#1064 unmerged PRs have the support for CLIs. 
The current plan is to first get  this PR merged. @shyam77git Please review and approve this PR. 
Then rework changes in #1064 PRs to support these design updates and submit to community for coming release.

**Comment by zhangyanzhao** [comment]
merge this with reviewers approval

**Comment by zhangyanzhao** [comment]
@bhaveshdell  can you please help to update the code PRs for this HLD by referring to #806 ? So that we can track this feature completeness clearly. Thanks.

**Comment by stephenxs** [comment]
Hi @bhaveshdell 
Looks like the corresponding PRs are still tracked in a closed PR https://github.com/sonic-net/SONiC/pull/761 which is confusing.
Could you please track the code PRs in this HLD which is latest?
Thanks

**Comment by venkatmahalingam** [comment]
> Hi @bhaveshdell Looks like the corresponding PRs are still tracked in a closed PR #761 which is confusing. Could you please track the code PRs in this HLD which is latest? Thanks

@stephenxs Yes, we'll add code PRs to this HLD.

**Comment by shyam77git** [comment]
> > Hi @bhaveshdell Looks like the corresponding PRs are still tracked in a closed PR #761 which is confusing. Could you please track the code PRs in this HLD which is latest? Thanks
> 
> @stephenxs Yes, we'll add code PRs to this HLD.

Thanks @venkatmahalingam . 
That would help reviewers as to what all code PRs to be reviewed in context of this HLD.
Also, would it there be some new code PRs related to updates in this HLD?
@bhaveshdell  - it would be good if you can provide a full view here w.r.t code PRs (existing, new/upcoming etc.)

**Comment by shyam77git** [comment]
@bhaveshdell 
Based on last comment, could you please update the latest and further plan on this?
Please add code PRs list to this HLD for review

**Comment by zhangyanzhao** [comment]
Copy the code PRs from https://github.com/sonic-net/SONiC/pull/761

Corresponding Coding PRs:
https://github.com/Azure/sonic-buildimage/pull/7813
https://github.com/Azure/sonic-swss-common/pull/490
https://github.com/Azure/sonic-mgmt-framework/pull/85
https://github.com/Azure/sonic-mgmt-common/pull/48

**Comment by zhangyanzhao** [comment]
> The event/alarm framework HLD was reviewed and committed earlier. With the availability of event producer framework, the design is updated to provide DB/persistent support for event history. Also gnmi/rest openconfig interface for get operations and gnmi subscription. Code PRs copied from
> 
> [sonic-net/sonic-buildimage#7813](https://github.com/sonic-net/sonic-buildimage/pull/7813) [sonic-net/sonic-swss-common#490](https://github.com/sonic-net/sonic-swss-common/pull/490) [sonic-net/sonic-mgmt-framework#85](https://github.com/sonic-net/sonic-mgmt-framework/pull/85) [sonic-net/sonic-mgmt-common#48](https://github.com/sonic-net/sonic-mgmt-common/pull/48)

@venkatmahalingam  @bhaveshdell  can you please check if above code PRs are accurate or others should be added? I copied them from #761. Code PRs should be included in HLD per community  agreement. Thanks.

**Comment by zhangyanzhao** [comment]
one code PR is still open, move to backlog

**Comment by zhangyanzhao** [comment]
handle the reverted code with 202505 release.

**Pull request merged**